src/js/ripple/amount.js

```diff
@@ -574,40 +567,30 @@ Amount.prototype.parse_quality = function(quality, counterCurrency, counterIssue
   this._issuer      = UInt160.from_json(counterIssuer);
   this._is_native   = this._currency.is_native();
 
-  var power = 0;
-  if (this._is_native) {
-    if (opts.inverse) {
-      power += 1;
-    } else {
-      power -= 1;
-    }
+  if (this._is_native && baseCurrency.is_native()) {
+    throw new Error('XRP/XRP quality is not allowed');
   }
 
-  // Correct offset if xrp_as_drops option is not set and base currency is XRP
-  if (!opts.xrp_as_drops &&
-      baseCurrency.is_valid() &&
-      baseCurrency.is_native()) {
-    if (opts.inverse) {
-      power -= 1;
+  var adjusted;
+  var inversion = opts.inverse ? -1 : 1;
+  // Correct for XRP to drops factors
+  if (!opts.xrp_as_drops) {
```

Comment:
```javascript
  /*
  The quality, as stored in the last 64 bits of a directory index, is stored as
  the quotient of TakerPays/TakerGets. 

  When `opts.inverse` is true we are looking at a quality used for determining a
  `bid` price and it must first be inverted, before our declared base/counter
  currencies are in line with the price.

  For example:

    quality as stored :  5 USD          /  3000000 drops
    inverted          :  3000000 drops  /          5 USD
  */
  var adjusted = opts.inverse ? inverse(value) : value;
  var nativeAdjusted = adjusted;

  if (!opts.xrp_as_drops) {
    // `In a currency exchange, the exchange rate is quoted as the units of the
    //  counter currency in terms of a single unit of a base currency`. A
    //  quality is how much taker must `pay` to get ONE `gets` unit thus:
    //    pays ~= counterCurrency 
    //    gets ~= baseCurrency.
    if (this._is_native) {
      // pay:$price              drops  get:1 X
      // pay:($price / 1,000,000)  XRP  get:1 X
      nativeAdjusted = adjusted.div(Amount.bi_xns_unit);
    } else if (baseCurrency.is_valid() && baseCurrency.is_native()) {
      // pay:$price X                   get:1 drop
      // pay:($price * 1,000,000) X     get:1 XRP
      nativeAdjusted = adjusted.times(Amount.bi_xns_unit); 
    }
  }
  
  this._set_value(nativeAdjusted);
```

I still prefer this form, as I find it easier to read. That's debatable of course, but it uses measurably less steps.